VideoPlayer: VAAPI - Add av1 hw decoding

a1rwulf opened the pull request:
## Description
Add AV1 hw decoding via VAAPI.

## Motivation and context
I've got an Intel TGL system in hands, so thought I can give it a shot.

## How has this been tested?
Tested on Ubuntu 21.04.

## Screenshots (if appropriate):
![image](https://user-images.githubusercontent.com/5141257/116747657-85697380-a9fe-11eb-8688-01e02c5cd552.png)


## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [x] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

As you can see it's not quite working yet.
But good for review and I'd like to get input on what might go wrong.

fritsch commented:
Can you dump the values you put into the EGL-Texture?
8 additional frames are enough according to the ffmpeg commit, like in VP9 case.

lrusak commented on xbmc/cores/VideoPlayer/DVDCodecs/Video/VAAPI.cpp:
```diff
@@ -670,6 +672,18 @@ bool CDecoder::Open(AVCodecContext* avctx, AVCodecContext* mainctx, const enum A
+    case AV_CODEC_ID_AV1:
+    {
+      if (avctx->profile == FF_PROFILE_AV1_MAIN)
+        profile = VAProfileAV1Profile0;
```
Might be worth checking what version of libva this was added in as we might have to bump the minimum version of add guards

bkuhls replied:
av1 decoding was added with libva-2.7.0:
https://github.com/intel/libva/blob/master/NEWS#L77
https://github.com/intel/libva/commit/7eb84e41658944ab0713c5aa6b8eddbcb50a114d

a1rwulf replied:
I used the same VA_CHECK_VERSION that is used inside ffmpeg, which is 1.8.0.

a1rwulf replied:
Interestingly those defines seem to work for our freebsd builder, which complained about missing AV1 defines and works now.
I'll have a closer look once again.

jernejsk replied:
sorry, I took another look. API version is disconnected from lib version. lib 2.7.0 has API version 1.7.0. I guess ffmpeg depends on some later fixes to require API version 1.8.0. I think this PR is ready.

a1rwulf replied:
Yeah just confirmed the same.

a1rwulf commented:
Seems like mpv has similar issues with the hw decoder: https://github.com/intel/media-driver/issues/1155

a1rwulf commented:
https://github.com/intel-media-ci/ffmpeg/pull/403 fixes the artifacts issue (at least for 8bit material).

> 8 additional frames are enough according to the ffmpeg commit, like in VP9 case.

That's not the case, I still had artifacts on reference frames that disappear once I set the value to 16 again.

a1rwulf commented:
![Uploading IMG_20210515_153137.jpg…]()

a1rwulf commented:
![IMG_20210515_153137](https://user-images.githubusercontent.com/5141257/118363316-11010980-b594-11eb-844a-b5c290ad4af4.jpg)

fritsch commented:
> [intel-media-ci/ffmpeg#403](https://github.com/intel-media-ci/ffmpeg/pull/403) fixes the artifacts issue (at least for 8bit material).
> 
> > 8 additional frames are enough according to the ffmpeg commit, like in VP9 case.
> 
> That's not the case, I still had artifacts on reference frames that disappear once I set the value to 16 again.

Okay - was worth a try. See for 4K it makes a huge difference ~ 100 MB.

a1rwulf commented:
I added the libva version check and added a way to select av1 hw decoders and fallback to libdav1d.
Not sure how clean this is, I see other projects iterating over all codecs and select hwaccels etc. but I don't really want to refactor our whole logic, to me the chosen approach doesn't look too bad unless I miss something.

heitbaum commented:
so I have tested this on my dev tree, which is Kodi 19.1, https://github.com/xbmc/FFmpeg/releases/tag/4.4-N-Alpha1, LE10+, 5.12.4, GuC=2/HuC, NUC11 with TGL i7-1165G7. 

Results below for 8bit. 10bit AV1s just core dump using the HW decoder.

![F58AB2ED-82DA-4039-8BA1-67249F432CD3](https://user-images.githubusercontent.com/6086324/118631700-1c159d00-b813-11eb-84eb-dfef5ff5b91d.jpeg)
![0FA145AB-3AE7-4424-8B57-997527E105CE](https://user-images.githubusercontent.com/6086324/118631718-1fa92400-b813-11eb-93ad-2b1bb8548cc8.jpeg)
![11B23A6B-35AA-46C5-A17A-52C28920093B](https://user-images.githubusercontent.com/6086324/118631730-20da5100-b813-11eb-87cb-1d015cae0123.jpeg)
![A82F389E-EFC8-47FC-B8D6-00B2A8EB1D5C](https://user-images.githubusercontent.com/6086324/118631736-2172e780-b813-11eb-80f3-a5f4f76d84ed.jpeg)


For the 10bits - detail below:

Core dumped in /usr/lib/libavcodec.so.58 - 10bit HW decode is not working for AV1. http://paste.ubuntu.com/p/DR6fqDHd8q/ 

Tested with both Chimera-2397fps-AV1-10bit-1920x1080-3365kbps.mkv and Chimera-2397fps-AV1-10bit-3840x2160-12407kbps.mkv . 

If I software decode (both play): 
- 100% CPU on the HD and 300% on the 4K
- GPU @ 27% for HD and 60% for 4K. 
- CPU is i7-1165G7. 
- The 4K SW decode is jittery

lrusak commented:
@heitbaum That might be something that needs to be reported to the intel people. I'm not even sure if 10bit is supported at this time.

a1rwulf commented:
> @heitbaum That might be something that needs to be reported to the intel people. I'm not even sure if 10bit is supported at this time.

I've done that already

a1rwulf commented:
Intel will provide some improvements for 10bit AV1: https://github.com/intel-media-ci/ffmpeg/pull/412
While it works for mpv, we still crash - so I need to look into that.
Once the 10bit crash is resolved and the patches hit 4.4 ffmpeg, I assume we can go forward with merging.

xuguangxin commented:
@a1rwulf @lrusak , good to know you will start av1 activities.
You can report an issue to FFmpeg or just @ us for help. 
We will take it as a high priority.

thanks

rmkimathi commented:
What is the status of this AV1 hardware merge into KODI? ETA

basilgello commented:
@rmkimathi The merge makes sense when it is proven to work :) At the moment, 10bit is broken at least. So no ETAs :)

xuguangxin commented:
@basilgello , anything we can help with?

basilgello commented:
> @basilgello , anything we can help with?

@a1rwulf ?

a1rwulf commented:
@xuguangxin I'm not sure yet, haven't had time to investigate it in more depth.
I see a crash in kodi that I do not see with mpv, so I assume we use ffmpeg a bit differently and it might be a kodi issue.

xuguangxin commented:
@a1rwulf , thanks.
how about we let 8 bits online, then we fix the 10 bits later. 
Not every clip is 10 bits clip
thanks

a1rwulf commented:
@xuguangxin sorry for my late reply, I'm pretty busy with life atm so not sure when I can either fix or disable hw accel for 10-bit.
I'll try to discuss it inside the team what to do with this.

xuguangxin commented:
@a1rwulf , sure. Do not hesitate to ping us for anything if you think we can help.

a1rwulf commented:
@xuguangxin I'm back to try fix the 10bit crash I see.
Maybe someone of your team that is familiar with ffmpeg, can make sense out of following backtrace:
```
skip_mode_params (s=0x7fff740337c0) at src/libavcodec/av1dec.c:268
268	        ref_hint = s->ref[header->ref_frame_idx[i]].raw_frame_header->order_hint;

#0  skip_mode_params (s=0x7fff740337c0) at src/libavcodec/av1dec.c:268
#1  0x00007ffff5cfc193 in get_current_frame (avctx=0x7fff740c7e00) at src/libavcodec/av1dec.c:1017
#2  0x00007ffff5cfc6a8 in av1_decode_frame (avctx=0x7fff740c7e00, frame=0x7fff74157900, got_frame=0x7fff5a2fbff8, pkt=0x7fff74049900) at src/libavcodec/av1dec.c:1140
#3  0x00007ffff5dfafdc in decode_simple_internal (avctx=0x7fff740c7e00, frame=0x7fff74157900, discarded_samples=0x7fff5a2fc0b0) at src/libavcodec/decode.c:329
#4  0x00007ffff5dfbca5 in decode_simple_receive_frame (avctx=0x7fff740c7e00, frame=0x7fff74157900) at src/libavcodec/decode.c:530
#5  0x00007ffff5dfbda3 in decode_receive_frame_internal (avctx=0x7fff740c7e00, frame=0x7fff74157900) at src/libavcodec/decode.c:550
#6  0x00007ffff5dfc087 in avcodec_send_packet (avctx=0x7fff740c7e00, avpkt=0x7fff50011540) at src/libavcodec/decode.c:617
#7  0x000055555617c49e in CDVDVideoCodecFFmpeg::AddData(DemuxPacket const&) ()
#8  0x000055555622c813 in CVideoPlayerVideo::Process() ()
#9  0x000055555650cc45 in CThread::Action() ()
#10 0x000055555650ee62 in CThread::Create(bool)::{lambda(CThread*, std::promise<bool>)#1}::operator()(CThread*, std::promise<bool>) const [clone .constprop.0] ()
#11 0x000055555650f85a in std::thread::_State_impl<std::thread::_Invoker<std::tuple<CThread::Create(bool)::{lambda(CThread*, std::promise<bool>)#1}, CThread*, std::promise<bool> > > >::_M_run() ()
#12 0x00007ffff3b4b694 in  () at /lib/x86_64-linux-gnu/libstdc++.so.6
#13 0x00007ffff38ef450 in start_thread (arg=0x7fff5a2fd640) at pthread_create.c:473
#14 0x00007ffff3da8d53 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

```
(gdb) p s->ref[header->ref_frame_idx[i]]
$2 = {tf = {f = 0x7fff741b78c0, owner = {0x0, 0x0}, progress = 0x0}, tf_display = {f = 0x7fff74092900, owner = {0x0, 0x0}, progress = 0x0}, hwaccel_priv_buf = 0x0, hwaccel_picture_private = 0x0, header_ref = 0x0, 
  raw_frame_header = 0x0, temporal_id = 0, spatial_id = 0, gm_invalid = "\000\000\000\000\000\000\000", gm_type = "\000\000\000\000\000\000\000", gm_params = {{0, 0, 0, 0, 0, 0}, {0, 0, 0, 0, 0, 0}, {0, 0, 0, 0, 0, 0}, {0, 0, 
      0, 0, 0, 0}, {0, 0, 0, 0, 0, 0}, {0, 0, 0, 0, 0, 0}, {0, 0, 0, 0, 0, 0}, {0, 0, 0, 0, 0, 0}}, skip_mode_frame_idx = "\000", film_grain = {apply_grain = 0 '\000', grain_seed = 0, update_grain = 0 '\000', 
    film_grain_params_ref_idx = 0 '\000', num_y_points = 0 '\000', point_y_value = '\000' <repeats 13 times>, point_y_scaling = '\000' <repeats 13 times>, chroma_scaling_from_luma = 0 '\000', num_cb_points = 0 '\000', 
    point_cb_value = "\000\000\000\000\000\000\000\000\000", point_cb_scaling = "\000\000\000\000\000\000\000\000\000", num_cr_points = 0 '\000', point_cr_value = "\000\000\000\000\000\000\000\000\000", 
    point_cr_scaling = "\000\000\000\000\000\000\000\000\000", grain_scaling_minus_8 = 0 '\000', ar_coeff_lag = 0 '\000', ar_coeffs_y_plus_128 = '\000' <repeats 23 times>, ar_coeffs_cb_plus_128 = '\000' <repeats 24 times>, 
    ar_coeffs_cr_plus_128 = '\000' <repeats 24 times>, ar_coeff_shift_minus_6 = 0 '\000', grain_scale_shift = 0 '\000', cb_mult = 0 '\000', cb_luma_mult = 0 '\000', cb_offset = 0, cr_mult = 0 '\000', cr_luma_mult = 0 '\000', 
    cr_offset = 0, overlap_flag = 0 '\000', clip_to_restricted_range = 0 '\000'}, coded_lossless = 0 '\000'}
```

raw_frame_header is 0x0 for me, but I have no idea what that means.

a1rwulf commented:
Also here are logfiles for 8bit and 10bit - also ffmpeg logs an error: `[av1] Failed to end picture decode issue: 18 (invalid parameter).`
[kodi_av1_8bit.log](https://github.com/xbmc/xbmc/files/7659427/kodi_av1_8bit.log)
[kodi_av1_10bit.log](https://github.com/xbmc/xbmc/files/7659430/kodi_av1_10bit.log)

xuguangxin commented:
Welcome back. 
@feiwan1 could you help check this

feiwan1 commented:
Hi @a1rwulf, could you share me your test clip? All my clips can't reproduce your issue.

a1rwulf commented:
Hi @feiwan1, http://download.opencontent.netflix.com.s3.amazonaws.com/AV1/Chimera/Old/Chimera-AV1-10bit-1920x1080-6191kbps.mp4

It's this sample I shared in the original media-driver ticket.
How did you try to reproduce?
The thing is, it must be something in Kodi's usage of ffmpeg, as it works in mpv - I tried to crosscheck with the mpv implementation but I can't see any obvious issue in our implementation so far that would cause it.

Also for easier communictation, is there some irc channel where we can communicate?
If not, we have a kodi channel on libera.chat you can ping me there.

feiwan1 commented:
> Hi @feiwan1, http://download.opencontent.netflix.com.s3.amazonaws.com/AV1/Chimera/Old/Chimera-AV1-10bit-1920x1080-6191kbps.mp4
> 
> It's this sample I shared in the original media-driver ticket. How did you try to reproduce? The thing is, it must be something in Kodi's usage of ffmpeg, as it works in mpv - I tried to crosscheck with the mpv implementation but I can't see any obvious issue in our implementation so far that would cause it.
I tried to use ffmpeg vaapi cmdline to decode the clip, it can work too:
$ ffmpeg -hwaccel vaapi -hwaccel_device /dev/dri/renderD128 -v verbose  -i Chimera-AV1-10bit-1920x1080-6191kbps.mp4 -pix_fmt yuv420p10le -vsync passthrough -f framemd5 -y out.md5

From your BT message, it seems the reference list(s->ref[]) is broken. It records the reference frames which will be used to decode current frame, and updated(update_reference_list()) after current frame decoded. s->ref[].raw_frame_header is the reference frame information parsered from reference frame header. You can trace the reference list to see what happened when decode the 1st key frame before the crashed frame, normally s->ref[] will be all updated to the key frame after it is decoded, that mens all s->ref[].raw_frame_header should not be null after key frame is decoded succesfully.

> 
> Also for easier communictation, is there some irc channel where we can communicate? If not, we have a kodi channel on libera.chat you can ping me there.
Sure, we can talk in kodi channel if needed. My nick name is feiw there.

feiwan1 commented:
Hi @a1rwulf, how is this issue now? If it is still exist, could you share me reproduce steps. I can help try to set it up and check it in my side.

a1rwulf commented:
hi @feiwan1,
After updating the whole stack (kodi, ffmpeg, mesa, libva, mediadriver) to latest versions, it still crashes for me. Unfortunately I got stuck after your message about how to approach to debug ffmpeg, due to a wrist surgery so I'm somewhat handycapped at the moment.
Reproduction of the issue is easy, just save above sample and start playback it will immediately crash.

If you need help to setup the environment, I can definitely help there.
Not sure what OS etc. you are on, but here are our docs:
https://github.com/xbmc/xbmc/tree/master/docs

a1rwulf commented:
@feiwan1 thank you for your help regarding the crash - decoding 10bit material works now :-)
It's still a bit WIP, I need to verify if the taken approach to transport the bitdepth is considered OK.
Seems the `bits_per_raw_sample` field is meant to hold that information, but the only reliable way to retrieve it is by taking it from the pix_fmt.
So I still use the field to carry the information into the VAAPI class.

a1rwulf commented:
@lrusak How do we deal with ffmpeg updates?
The required patches for ffmpeg have been applied to upstream ffmpeg 4.4.

lrusak commented:
> @lrusak How do we deal with ffmpeg updates? The required patches for ffmpeg have been applied to upstream ffmpeg 4.4.

Make a PR to https://github.com/xbmc/ffmpeg/tree/release/4.4-kodi and if we merge it then we can make a new release and bump the ffmpeg version here.

a1rwulf commented:
> The required patches for ffmpeg have been applied to upstream ffmpeg 4.4.

Appearently I dreamed above, the patches have been merged into ffmpeg's master.
Not sure where they will land and how to find it out...

a1rwulf committed:
VideoPlayer: VAAPI - Add av1 hw decoding

a1rwulf commented:
Needs https://github.com/xbmc/FFmpeg/pull/22 for proper playback of 10bit AV1 content.
Not sure why windows fails due to ffmpeg not found, FreeBSD seems to need a more recent libva (I added the ifdef for now).

Also after I had a chat with the ffmpeg guys, I use the pix_fmt to determine the bitdepth and also the HEVC10bit profile can still contain 8bit video, so we only chose the HEVC 10bit format if it's really 10bit video.

a1rwulf commented:
Code is finished from my side, so comments would be appreciated.
Cannot be merged until we rebased our ffmpeg verison on top of upstream 4.4.

xuguangxin commented:
Great!
@a1rwulf  do you have ETA for the 4.4 rebase. 
Any other issue we can help with?
thanks

a1rwulf commented:
Thanks @lrusak for rebasing our ffmpeg ontop of upstream 4.4.1.
I PR'ed the AV1 patches: https://github.com/xbmc/FFmpeg/pull/23
Now that we settled all necessary parts, only review and merges are pending to get done with this :-)

lrusak commented:
@a1rwulf can you add the ffmpeg bump to this PR?

https://github.com/xbmc/FFmpeg/releases/tag/4.4.1-Nexus-Alpha1

a1rwulf committed:
Bump ffmpeg version to 4.4.1

a1rwulf committed:
DVDStreamInfo: Add bitdepth property

thexai commented:
Is needed refresh ffmpeg Windows patches (https://github.com/xbmc/xbmc/tree/master/tools/buildsteps/windows/patches) prior to merge.

a1rwulf commented:
> Is needed refresh ffmpeg Windows patches (https://github.com/xbmc/xbmc/tree/master/tools/buildsteps/windows/patches) prior to merge.

I don't have a windows setup, would you be able to help me out with this @thexai ?

lrusak commented:
> > Is needed refresh ffmpeg Windows patches (https://github.com/xbmc/xbmc/tree/master/tools/buildsteps/windows/patches) prior to merge.
> 
> 
> 
> I don't have a windows setup, would you be able to help me out with this @thexai ?

You can just apply them to the ffmpeg tree. They might only need some small adjustments.

a1rwulf commented:
```
16:26:08 [1mffmpeg-4.4.1-Nexus-Alpha1[0;10m ..................................... [[34mExtracting[0;10m]
16:26:12 Applying patch /xbmc/tools/buildsteps/windows/patches/0001-ffmpeg-windows-configure-detect-openssl.patch
16:26:12 (Stripping trailing CRs from patch; use --binary to disable.)
16:26:12 patching file configure
16:26:12 Hunk #1 succeeded at 6530 (offset 1 line).
16:26:12 Applying patch /xbmc/tools/buildsteps/windows/patches/0002-ffmpeg-windows-configure-fix-zlib-conflict.patch
16:26:12 (Stripping trailing CRs from patch; use --binary to disable.)
16:26:12 patching file configure
16:26:12 Applying patch /xbmc/tools/buildsteps/windows/patches/0003-ffmpeg-windows-configure-allow-building-static.patch
16:26:12 (Stripping trailing CRs from patch; use --binary to disable.)
16:26:12 patching file configure
16:26:12 Hunk #1 succeeded at 5440 (offset 1 line).
16:26:12 Hunk #2 succeeded at 5491 (offset 1 line).
16:26:12 Applying patch /xbmc/tools/buildsteps/windows/patches/0004-ffmpeg-windows-configure-detect-libdav1d.patch
16:26:12 (Stripping trailing CRs from patch; use --binary to disable.)
16:26:12 patching file configure
16:26:12 Hunk #1 succeeded at 6358 (offset 1 line).
16:26:12 [1mffmpeg-4.4.1-Nexus-Alpha1 (win32)[0;10m ............................ [[34mConfiguring[0;10m]
```

Looks like they still apply fine.

fuzzard commented:
The intent would be to have no offsets/fuzzing matches, so the patches apply cleanly on current release. The more it gets pushed to latert, the more effort it becomes later

a1rwulf commented:
> The intent would be to have no offsets/fuzzing matches, so the patches apply cleanly on current release. The more it gets pushed to latert, the more effort it becomes later

No worries, I can do that.

However I'm a bit puzzled by this one:
```
C:/Users/Kodi/Kodi/jenkins/builds/workspace/WIN-32/project/BuildDependencies/build/src/ffmpeg-293e067b0c/libavcodec/libdav1d.c(151): fatal error C1017: invalid integer constant expression
```

a1rwulf commented:
Jenkins build this please

a1rwulf committed:
Depends: Bump dav1d to 0.8.2

a1rwulf commented:
Retry to see if it still pulls the old dav1d: Jenkins build this please

a1rwulf commented:
Alright jenkins is happy.
@lrusak Would you mind having a look at the code?

lrusak commented on xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecFFmpeg.cpp:
```diff
@@ -336,7 +336,16 @@ bool CDVDVideoCodecFFmpeg::Open(CDVDStreamInfo &hints, CDVDCodecOptions &options
+  // The decoder state check is needed to succesfully fallback to sw decoding if
+  // necessary.
+  if (hints.codec == AV_CODEC_ID_AV1 && m_decoderState != STATE_HW_FAILED)
+    pCodec = avcodec_find_decoder_by_name("av1");
```
This seems strange to have to workaround this. Are you sure the dav1d decoder actually gets selected here? The hwaccell should get setup in the `get_format` callback. Otherwise I would expect this to be the same behaviour for x264, etc software decoders.

a1rwulf replied:
As described in the comment:
The dav1d decoder is a standalone decoder in ffmpegs decoder list, which appears to be inserted int he codec list before the decoder named `av1` (which is libaom/hwaccell).
If you look at `avcodec_find_decoder` it just looks at the `codec_id` and returns the first possible decoder which in turn will always be dav1d as soon as it's enabled in ffmpeg.

This is not true for h264 and hevc, because the x264/x265 decoders are not pushed into the decoder list as standalone decoders.
You can also see this when you use the command `ffmpeg -decoders`:

```
 V..... libdav1d             dav1d AV1 decoder by VideoLAN (codec av1)
 V....D av1                  Alliance for Open Media AV1
VFS..D h264                 H.264 / AVC / MPEG-4 AVC / MPEG-4 part 10
 V..... h264_v4l2m2m         V4L2 mem2mem H.264 decoder wrapper (codec h264)
 VFS..D hevc                 HEVC (High Efficiency Video Coding)
 V..... hevc_v4l2m2m         V4L2 mem2mem HEVC decoder wrapper (codec hevc)
```

Seems the reason is that there are 2 possible sw decoders dav1d and libAOM and the hwaccell is bound to "av1".
Other projects like mpv select codecs by iterating over codecs with hwaccells so they deal that stuff differently, but honesty I don't want to rework our whole codec selection code, hence I came up with this.

lrusak commented on xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp:
```diff
@@ -1634,6 +1639,11 @@ CDemuxStream* CDVDDemuxFFmpeg::AddStream(int streamIdx)
+        const AVPixFmtDescriptor* desc =
+            av_pix_fmt_desc_get(static_cast<AVPixelFormat>(pStream->codecpar->format));
+        if (desc != nullptr && desc->comp != nullptr)
+          st->bitDepth = desc->comp[0].depth;
```
Should we be using `av_get_bits_per_pixel` here?

a1rwulf replied:
Yep, seems to do about the same thing - will change.

a1rwulf replied:
OK the additional computation of this method does not give back what I'm looking for.
For 10 bit AV1 it returns 15, while my approach returns 10 - so I'll leave it as is.

fritsch replied:
What does the mediainfo of this file say? 15 is odd in all combinations.

fritsch replied:
Ah, obviously. It's YUV with a YUV 4:2:0 subsamling. That means: per 4 pixels: 4*10+10+10 = 60 -> 60/4 = 15.

All fine here. You don't want the bits_per_pixel, but the depth I assume.

fritsch replied:
Your original approach was okay I think, also get_bits_depth is going to be deprecated. https://patchwork.ffmpeg.org/project/ffmpeg/patch/20210822145409.316517-2-george@nsup.org/ as well and the fallback is using something similar.

lrusak reviewed (commented):
Looks good! It will be nice to get this change in. I've left a couple comments about some things I saw. Hopefully we can resolve them quickly and get this in.

lrusak commented:
Can you squash the fixup and reorder some of the commits to be in a logic build order (as best as possible). The dav1d bump commit should probably go first, etc.

a1rwulf commented:
Jenkins build this please

lrusak commented:
@a1rwulf youll have to update the sha512 hash now that #20864 has merged.

You can do that here: https://github.com/xbmc/xbmc/blob/28bb4b09387d7a8e2ea799ef15f2bce4b8b6f1db/tools/depends/target/ffmpeg/FFMPEG-VERSION#L5

a1rwulf commented:
> @a1rwulf you'll have to update the sha512 hash now that #20864 has merged.
> 
> You can do that here:
> 
> https://github.com/xbmc/xbmc/blob/28bb4b09387d7a8e2ea799ef15f2bce4b8b6f1db/tools/depends/target/ffmpeg/FFMPEG-VERSION#L5

Ah thanks, I didn't get a conflict for some reason and missed this entire change...

a1rwulf commented:
Finally all green!

rmkimathi commented:
### Kodi Media Center 20.0-ALPHA1 (19.90.101) Git:20220207-ca78545b83
ff-av1-vaapi (HW) now working. Key milestone worthy of celebration! Kodi v20 (Nexus) is now Release Candidate material.

![av1-select](https://user-images.githubusercontent.com/13484308/152769210-9fc88513-dacd-4939-8be8-6fcaef3d85bd.png)
